packages/foam-core/src/model/workspace.ts

```diff
@@ -0,0 +1,401 @@
+import { diff } from 'fast-array-diff';
+import { isEqual } from 'lodash';
+import * as path from 'path';
+import { URI } from '../common/uri';
+import { Resource, NoteLink, Note } from '../model/note';
+import {
+  computeRelativeURI,
+  isSome,
+  isNone,
+  parseUri,
+  placeholderUri,
+  isPlaceholder,
+} from '../utils';
+import { Emitter } from '../common/event';
+import { IDisposable } from 'index';
+
+export type Connection = {
+  source: URI;
+  target: URI;
+};
+
+export function getReferenceType(
+  reference: URI | string
+): 'uri' | 'absolute-path' | 'relative-path' | 'key' {
+  if (URI.isUri(reference)) {
+    return 'uri';
+  }
+  const isPath = reference.split('/').length > 1;
+  if (!isPath) {
+    return 'key';
+  }
+  const isAbsPath = isPath && reference.startsWith('/');
+  return isAbsPath ? 'absolute-path' : 'relative-path';
+}
+
+function normalizePath(pathValue: string) {
+  const { ext } = path.parse(pathValue);
+  return ext.length > 0 ? pathValue : pathValue + '.md';
+}
+
+function normalizeKey(pathValue: string) {
+  return path.parse(pathValue).name;
+}
+
+export class FoamWorkspace implements IDisposable {
```

Comment:
So I compared this `FoamWorkspace` implementation with the `NoteGraph` one and those are my observations - you might have to take them with a grain of salt as it's my first time looking at those.

First, `FoamWorkspace` was easier to understand, so good job on that!

Second I'm not sure I understand the reason for having static methods here. The static methods all take a `FoamWorkspace` instance as an argument, so why not have the instance methods directly perform the role of the static methods?

I see having the graph state in hashmaps as interesting, the `graphlib` dependency did seem overkill. I don't think Foam will need to perform complex or multi hop queries so it does seem fine and easier to understand like this. Maybe it'd be interesting to benchmark the two implementations? :thinking: Though I don't expect a huge difference. 

What I'm wondering about is: with a key value store implementation like this, can we guarantee ACID properties? and how does it handle concurrent operations? Maybe it's not that important as since it's an extension, the graph database should only have on client :thinking:

Reply:
Ah yeah, and I didn't get what you meant with:

> add plugin system to `FoamWorkspace`

?

Reply:
currently `bootstrap` uses the `NoteGraph`, and provides hooks for local plugins to augment the note graph. That needs to be ported to the `FoamWorkspace` API. see https://github.com/foambubble/foam/blob/master/docs/features/foam-local-plugins.md

Reply:
Thanks for the link, it makes sense now :+1:

Reply:
I actually missed your previous comment, let me reply.

> First, FoamWorkspace was easier to understand, so good job on that!

I really think NoteGraph is hard to understand, and graphlib was doing some magic that was creating more issues than it was solving. I am glad in that regard it is a success :)

> Second I'm not sure I understand the reason for having static methods here

A few reasons:
- I prefer the functional approach, I think it's easier to read in general (even though this is not pure functional because data is not immutable for performance reasons, I am "treating" it as if it were in the API, which still gives the readability benefit)
- A functional-only approach is a bit heavy imo when doing several operations (e.g. look at the tests when I set up the workspace). The class both provides a namespace for the operation, and a coherent grouping (I think it helps to think about the API)
- I have seen this pattern in the VSCode codebase (see `uri.ts`); mine is a bit stricter (all public fn have both static and instance methods) but the idea is the same (at times one approach works better than the other)
- I'd say the overhead of doing this hybrid approach is low (it would have actually been very low if typescript were a bit smarter, my original approach was to have all instance methods like `getNote = FoamWorkspace.getNote.bind(null, this)`, which is semantically correct but TS struggles to give me the proper type inferences when I later call that method :( )

Still, I am open to other views, I can see it's heavier than a pure OO or FP approach

> Maybe it'd be interesting to benchmark the two implementations? 

Yes I really should invest a bit of time creating some benchmarks. Both for simple in-memory operations, but also more real life e2e (generate a big repo with various types of connections)

> can we guarantee ACID properties?

node is single-threaded so I would expect that to just work? maybe I am missing something though (it's been a while since I have done serious multithreading)
also, I can see how some batch updates (e.g. rename a directory, which affects many notes) might require some special handling (but no different I think than when using graphlib)

Reply:
IMO most of this discussion would be solved with adhere to VSCode API style whenever possible, this makes it consistent and easier for people to pick since the documentation examples match the code.

Regarding the ACID properties, I am unsure why that is a concern. Regardless, once again, we should take advantage of the offered APIs instead of building our own from the ground up.

Reply:
> A few reasons

Thanks for the explanation. I see it interesting & working, just not used to seeing this

> Yes I really should invest a bit of time creating some benchmarks

I was thinking it's be simple to use an existing knowledge base for benchmarking; there are plenty of them out there that are public, huge, and using wikilinks. 

---

On another note, have you considered using the [workspace state](https://code.visualstudio.com/api/references/vscode-api#ExtensionContext) as a key value store instead of having it directly in the class (probably this change is out of scope here)? I understand there is separation of concern between foam-core and foam-vscode, but maybe a reference to it could be passed to `FoamWorkspace`.

This way Foam could persist the data between sessions. It wouldn't have to recompute the whole graph on startup. I'm not 100% sure whether VS Code puts the storage in memory though, so maybe there are performance concerns.

Another big change I can see is that it'd make any write operation asynchronous (see `update` on [memento stores](https://code.visualstudio.com/api/references/vscode-api#Memento)).

Reply:
interesting points! definitely worth investigating.
I am not familiar with public knowledge bases that could work for this, what did you have in mind?

Reply:
@riccardoferretti I was thinking it'd be fine to pick someone's repo, no? Here are some public ones: https://github.com/KasperZutterman/Second-Brain